### ISSUE TITLE
Modifying parameters of FSDP-wrapped module by hand without summon_full_params context

### ISSUE THREAD

**Issue by blunt-octopus** [pytorch/pytorch#117742]
# Modifying parameters of FSDP-wrapped module by hand without summon_full_params context

## Issue description

I am training a large language module using FSDP. 

I want to store EMA weights while training, and use the EMA-model during evaluation. To this end, I initialize EMA-model as a deepcopy of the original one, and then update its parameters after every batch by hand. (I do deepcopy *before* wrapping both models in FSDP).

Existing implementations seem to use FSDP.summon_full_params() context when updating parameters (see e.g. [here](https://github.com/pytorch/pytorch/blob/7b94d934313320eff91236540d37021d62bc8984/test/distributed/fsdp/test_fsdp_mixed_precision.py#L1353)). However, I do not have this option since my model is large enough that FSDP.summon_full_params() results in too high memory usage. 
In principle, it seems possible to update parameters of the EMA model directly, since both the original model and EMA model are stored identically. 
However, it turns out that after modifying EMA parameters, the output of `ema_model.forward()` is unchanged, suggesting that these modifications do not persist.

## Code example

```
...
model = MyModel()
ema_model = deepcopy(model)
ema_model.eval()
for param in ema_model.parameters():
                param.requires_grad_(False)
...
# here both models are wrapped in FSDP, resulting in identical representations

# in update_parameters() function:
...
self_param = itertools.chain(ema_model.parameters(), ema_model.buffers())
model_param = itertools.chain(model.parameters(), model.buffers())

for p_averaged, p_model in zip(self_param, model_param):
    with torch.no_grad():
        p_averaged.copy_(torch.lerp(p_averaged.detach(), p_model.detach(), 1 - self.ema_coefficient))
...

# during evaluation
...
ema_model.forward(batch) 
...
```

I apologize for not being able to provide a full code, since I use a framework where the relevant code is scattered (I use llm-foundry + composer [link](https://github.com/mosaicml/llm-foundry)).

The output of the last line does not change after running `update_parameters()`, although if I print out e.g. `p_averaged` when running `update_parameters()`, its value changes as expected.

Is there any way to update parameters of FSDP module by hand in this situation, without using `summon_full_params()` context?

## System Info
Result of running `collect_env.py` on my system is here:
https://gist.github.com/blunt-octopus/3edcbc83e639b2cdcad47ab0abcca623

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu @kwen2501

**Comment by awgu** [comment]
I agree that we should be able to do EMA with FSDP without `summon_full_params()` (as long as both the main and EMA copies both shard with FSDP identically).

There are 2 'gotchas' that might be happening:
1. The sharded parameters need to be exposed from `model.parameters()` when performing the EMA update. This mainly means that you should not run the EMA update after the main model's forward but before its backward. In other words, we should make sure to run the EMA update after backward or the optimizer step but before the next forward.
2. The EMA model did not free its all-gathered parameters, so the next forward does not re-all-gather and hence uses stale parameters.

For 2, we do not have a public API right now for how to force reshard the parameters, but I think the function in https://github.com/pytorch/pytorch/issues/117421#issuecomment-1890948734 should work.

**Comment by blunt-octopus** [comment]
Hello @awgu, thanks a lot for your comment.
I've tried to use the method you've linked (with `FSDP._runtime_utils._reshard()`). However, it does not seem to work properly.
As I understand it, the need for manual resharding arises from the fact that we use `ema_model.forward()` during validation, but do not ever use `ema_model.backward()`, which can lead to FSDP using pre-unsharded (and thus stale) weights. To address this problem, I put [manual resharding code](https://github.com/pytorch/pytorch/issues/117421#issuecomment-1890948734) in between validation and updating of EMA parameters. However, it fails with AssertionError "Expects storage to be allocated", occuring [here](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_flat_param.py#L2555), which I take to mean that the manual resharding was not needed? (I have also tried wrapping resharding inside `try: ... except` block, to avoid resharding the weghts when not needed, and found that it never executes without AssertionError at all).

Can you please elaborate on the conditions under which we should need manual resharding, or the reason for the above AssertionError?

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu @kwen2501

**Comment by awgu** [comment]
I am guessing that you might be hitting the `AssertionError` since you have a mix of FSDP modules that _have_ resharded after forward and ones that have not. 

On nightlies, the reshard code should become idempotent now (cc: @weifengpy), so if you have access to the nightlies, then the `AssertionError` should go away.

**Comment by mts42000** [comment]
any workarounds on this?

**Comment by awgu** [comment]
EMA on the sharded parameters should work fine with FSDP2. You can use the `FSDPModule.reshard()` method to ensure that parameters are resharded and registered as the sharded ones.

**Comment by amorehead** [comment]
@blunt-octopus, were you able to get an implementation of this up and running? If so, would you be open to sharing it with us all?

**Comment by WZMIAOMIAO** [comment]
I have implemented a method(unofficial) combining FSDP(`StateDictType`=`SHARDED_STATE_DICT`) and EMA.  I tested it on `PyTorch 2.1` and it works.

```python
import os
from typing import Dict, List
from collections import defaultdict

import torch
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP, StateDictType
import torch.distributed.checkpoint as dist_cp
from torch.distributed.checkpoint.default_planner import DefaultSavePlanner


class ShardEMAModel:
    def __init__(self, fsdp_model: FSDP, decay: float = 0.999):
        assert isinstance(fsdp_model, FSDP)

        self.fsdp_model = fsdp_model
        self.decay = decay
        self.shard_ema_state: Dict[str, List[torch.Tensor]] = defaultdict(list)

        shard_state = self._get_shard_state()
        for k, v in shard_state.items():
            for local_shard in v._local_shards:
                self.shard_ema_state[k].append(local_shard.tensor.clone())

        self.num_shard_params = sum([sum([t.numel() for t in v]) for v in self.shard_ema_state.values()])
        print(f"Shard EMA Model has {self.num_shard_params / 1e6:.3f}M params.")

    def _get_shard_state(self):
        with FSDP.state_dict_type(self.fsdp_model, StateDictType.SHARDED_STATE_DICT):
            shard_state = self.fsdp_model.state_dict()

        return shard_state

    @torch.inference_mode()
    def update(self):
        """update EMA Model shard weights"""
        shard_state = self._get_shard_state()
        for k, v in shard_state.items():
            for idx, local_shard in enumerate(v._local_shards):
                self.shard_ema_state[k][idx].mul_(self.decay).add_(local_shard.tensor, alpha=1 - self.decay)

    def save_ema_shard_weights(self, save_dir: str):
        """save EMA Model shard weights"""
        with FSDP.state_dict_type(self.fsdp_model, StateDictType.SHARDED_STATE_DICT):
            os.makedirs(save_dir, exist_ok=True)

            shard_state = self.fsdp_model.state_dict()
            for k, v in shard_state.items():
                for idx, local_shard in enumerate(v._local_shards):
                    local_shard.tensor = self.shard_ema_state[k][idx]

            state_dict = {"model": shard_state}
            dist_cp.save(
                state_dict=state_dict,
                storage_writer=dist_cp.FileSystemWriter(save_dir),
                planner=DefaultSavePlanner(),
            )

    def save_shard_weights(self, save_dir: str):
        """save original FSDP Model shard weights"""
        with FSDP.state_dict_type(self.fsdp_model, StateDictType.SHARDED_STATE_DICT):
            os.makedirs(save_dir, exist_ok=True)

            shard_state = self.fsdp_model.state_dict()
            state_dict = {"model": shard_state}
            dist_cp.save(
                state_dict=state_dict,
                storage_writer=dist_cp.FileSystemWriter(save_dir),
                planner=DefaultSavePlanner(),
            )

```

test code:
```python
# create FSDP Model and EMA Model
fsdp_model = FSDP(...)
ema_model = ShardEMAModel(fsdp_model, decay=0.99)

# train fsdp model and optimizer weights
...

# update EMA Model shard weights
ema_model.update()

# save EMA Model shard weights
ema_model.save_ema_shard_weights("save_path")
```
